File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-1.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> S;
        for (string t : tokens) {
            if (t == "+") {
                int a = S.top(); S.pop();
                int b = S.top(); S.pop();
                S.push(b + a);
            } else if (t == "-") {
                int a = S.top(); S.pop();
                int b = S.top(); S.pop();
                S.push(b - a);
            } else if (t == "*") {
                int a = S.top(); S.pop();
                int b = S.top(); S.pop();
                S.push(b * a);
            } else if (t == "/") {
                int a = S.top(); S.pop();
                int b = S.top(); S.pop();
                S.push(b / a);
            } else {
                S.push(stoi(t));
            }
        }
        return S.top();
    }
};

File: Data Structures & Algorithms/minimum-stack/submission-1.rs
struct MinStack {
    minStack: Vec<i32>,
    stack: Vec<i32>
}

impl MinStack {
    pub fn new() -> Self {
        MinStack {
            minStack: vec![],
            stack: vec![]
        }
    }

    pub fn push(&mut self, val: i32) {
        match self.minStack.last() {
            None => self.minStack.push(val),
            Some(&n) if n >= val => self.minStack.push(val),
            _ => ()
        }
        self.stack.push(val);
    }

    pub fn pop(&mut self) {
        if self.stack.pop() == self.minStack.last().copied() {
            self.minStack.pop();
        }
    }

    pub fn top(&self) -> i32 {
        match self.stack.last() {
            Some(&n) => n,
            None => -1
        }
    }

    pub fn get_min(&self) -> i32 {
        match self.minStack.last() {
            Some(&n) => n,
            None => -1
        }
    }
}
